tooling/hibernate-gradle-plugin/src/main/java/org/hibernate/orm/tooling/gradle/EnhancementHelper.java

```diff
@@ -36,8 +38,82 @@
  * @author Steve Ebersole
  */
 public class EnhancementHelper {
-	static void enhance(SourceSet sourceSet, EnhanceExtension options, Project project) {
-		final ClassLoader classLoader = toClassLoader( sourceSet.getRuntimeClasspath() );
+	public static void enhance(
+			DirectoryProperty classesDirectoryProperty,
+			InputChanges inputChanges,
+			EnhancementSpec enhancementDsl,
+			Project project) {
+		final Directory classesDir = classesDirectoryProperty.get();
+		final File classesDirFile = classesDir.getAsFile();
+
+		final Enhancer enhancer = generateEnhancer( classesDir, enhancementDsl );
+
+		final String classesDirPath = classesDirFile.getAbsolutePath();
+
+		inputChanges.getFileChanges( classesDirectoryProperty ).forEach(
+				change -> {
+					switch ( change.getChangeType() ) {
+						case ADDED:
+						case MODIFIED: {
+							final File changedFile = change.getFile();
+							if ( changedFile.getName().endsWith( ".class" ) ) {
+								final String classFilePath = changedFile.getAbsolutePath();
+								if ( classFilePath.startsWith( classesDirPath ) ) {
+									// we found the directory it came from
+									//		-use that to determine the class name
+									enhance( classesDirFile, changedFile, enhancer, project );
+									break;
+								}
+							}
+							break;
+						}
+						case REMOVED: {
+							// nothing to do
+							break;
+						}
+						default: {
+							throw new UnsupportedOperationException( "Unexpected ChangeType : " + change.getChangeType().name() );
+						}
+					}
+				}
+		);
+	}
+
+	private static void enhance(
+			File classesDir,
+			File javaClassFile,
+			Enhancer enhancer,
+			Project project) {
+		final byte[] enhancedBytecode = doEnhancement( classesDir, javaClassFile, enhancer );
+		if ( enhancedBytecode != null ) {
+			writeOutEnhancedClass( enhancedBytecode, javaClassFile, project.getLogger() );
+			project.getLogger().info( "Successfully enhanced class [" + javaClassFile.getAbsolutePath() + "]" );
+		}
+		else {
+			project.getLogger().info( "Skipping class [" + javaClassFile.getAbsolutePath() + "], not an entity nor embeddable" );
+		}
+	}
+
+	private static byte[] doEnhancement(File root, File javaClassFile, Enhancer enhancer) {
+		try {
+			final String className = determineClassName( root, javaClassFile );
+			final ByteArrayOutputStream originalBytes = new ByteArrayOutputStream();
+			try (final FileInputStream fileInputStream = new FileInputStream( javaClassFile )) {
+				byte[] buffer = new byte[1024];
+				int length;
+				while ( ( length = fileInputStream.read( buffer ) ) != -1 ) {
+					originalBytes.write( buffer, 0, length );
+				}
+			}
+			return enhancer.enhance( className, originalBytes.toByteArray() );
```

Comment:
We might well avoid reinventing wheels as follows:
```
return enhancer.enhance( className, Files.readAllBytes( javaClassFile.toPath() );
```
The util method was introduced from JDK7. Needless to say, it is more efficient and robust.

Reply:
Here yes you are correct.  NIO is not always faster nor more efficient though, so we need to stay away from that as a general code rule.

I've made this change